### PR TITLE
Start RileyMaze toolbox

### PR DESCRIPTION
**Pull request by MegFord**

### CODE REVIEW

File: data/RileyMaze/panel.ui
Comment (ptomato) [inline data/RileyMaze/panel.ui:17]: I would suggest giving the Section widgets a `border_width`. The spacing in this panel is a bit tight.

File: data/RileyMaze/icons/16x16/actions/powerup-symbolic.svg
Comment (ptomato) [inline data/RileyMaze/icons/16x16/actions/powerup-symbolic.svg:27]: Minor comment and please consider this optional, but if you do happen to re-save the icons in Inkscape, you can save some space by saving them as "Plain SVG" which doesn't store all of Inkscape's editor state in them.

File: data/RileyMaze/panel.ui
Comment (ptomato) [inline data/RileyMaze/panel.ui]: As a heads up, we will probably need to retheme these radio buttons with something orange. (It was a guideline in the other toolboxes that all the clickable or draggable items should have some sort of orange indicator on them, but we didn't have any radio buttons in the other toolboxes as far as I recall.)

Check with Robin if she has any rough recommendations for this or leftover specs from a previous iteration. But it's fine to do this in a followup task, we should not wait on that for merging this.

File: data/RileyMaze/panel.ui
Comment (ptomato) [inline data/RileyMaze/panel.ui]: It's probably OK to remove the Pad A and Pad B components. They look like they are pretty straightforward copies of Robot A and Robot B, so they should be pretty easy to recreate.

File: data/application.css
Comment (ptomato) [inline data/application.css]: This (and some other removals below) looks like it might be an error from rebasing?

File: src/RileyMaze/combinedTopic.js
Comment (ptomato) [inline src/RileyMaze/combinedTopic.js]: I would name this `RMZUnitsTopic`. I can't remember why I named `LSCombinedTopic` that way, but that just goes to show it's not a good name to emulate :stuck_out_tongue_closed_eyes:

File: src/RileyMaze/combinedTopic.js
Comment (ptomato) [inline src/RileyMaze/combinedTopic.js]: I would rename this to `bindGlobalModel`, to correspond with `unbindGlobalModel`. We intended to do that for Lightspeed and Fizzics as well, but since we introduced the split between global and per-level unbind methods as a critical last-minute bug fix for episode 2 (!) we did not want to risk renaming things that were not directly affected. In any case, we should use the new names going forward.

File: src/RileyMaze/combinedTopic.js
Comment (ptomato) [inline src/RileyMaze/combinedTopic.js]: Similarly, this should be called `_bindLevelModel`.

File: src/RileyMaze/combinedTopic.js
Comment (ptomato) [inline src/RileyMaze/combinedTopic.js]: For consistency, `_onActiveChanged`.

File: src/RileyMaze/combinedTopic.js
Comment (ptomato) [inline src/RileyMaze/combinedTopic.js]: In the Riley Maze API document these were described as being `'up'` and `'down'` (without the capital letters.)

However, since that was written we've seen playtesters (and Hack employees) stumble over the quote chars for strings. We should check with Randy whether we should really add more, maybe we could make these into constants so you can just type `robotADirection = up;` in the code view without the quotes.

File: src/RileyMaze/model.js
Comment (ptomato) [inline src/RileyMaze/model.js]: Score target can be deleted, I think.

File: src/RileyMaze/toolbox.js
Comment (ptomato) [inline src/RileyMaze/toolbox.js]: This code is probably OK to be uncommented even though this game state isn't set yet.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: Here's where the `riley` object should go.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: Same comments here about renaming `bindGlobal` and `_bindModel`.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js:239]: In a few cases, the instructions codeview should work differently from all the other codeviews and call `_updateCode` with erroneous code, in order to pass it back to the game. This happens in cases where the game should display a glitched instruction tile. Double check with Randy when this is supposed to happen. My understanding is — typo in instruction name; too many instructions; too few instructions. But maybe he wants to have glitches in the game whenever there is an error at all, like in the case of invalid JS code?

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: Should be RMZ for consistency

**Review by ptomato — CHANGES_REQUESTED** [review]
This is working quite well so far, I had a lot of fun testing it together with the game branch. I think the biggest thing missing is validation of the `riley` object and its methods.

File: data/RileyMaze/icons/16x16/actions/powerup-symbolic.svg
Comment (MegFord) [reply data/RileyMaze/icons/16x16/actions/powerup-symbolic.svg:27]: Yes, I'd forgotten about that, thanks! This is just a placeholder asset but I'll do that in the future when I have to edit in Inkscape :)

File: data/application.css
Comment (MegFord) [reply data/application.css]: I think this branch diverged from master a while ago and it's a conflict, yeah.

File: src/RileyMaze/combinedTopic.js
Comment (MegFord) [reply src/RileyMaze/combinedTopic.js]: Okay! I thought it was because it combined a codeview and widgets.

File: data/RileyMaze/panel.ui
Comment (ptomato) [reply data/RileyMaze/panel.ui:17]: Border width didn't quite do what I was expecting, sorry for the wrong suggestion. I should have explained what I was aiming for in the first place instead of suggesting a solution.

In Lightspeed, there's a certain amount of space between the border of the Section widget and the widgets inside it:
Screenshot from 2019-04-23 17-53-43
In the robot Section, there are only a few pixels of space between the widgets and the border:
Screenshot from 2019-04-23 17-54-04
(and now that I look closer, it looks like the dark background behind "Robot A" doesn't extend all the way to the right side of the background box.)

I guess border width is not the right way to achieve that open spacing, but maybe take a look at Lightspeed to see what's different between these two widgets?

File: src/RileyMaze/RMZUnit.js
Comment (ptomato) [inline src/RileyMaze/RMZUnit.js:131]: For `up` and `down` to work as variables instead of string constants, they will have to be defined in the scope object. Currently if I choose "up" or "down" in the UI and let the code be regenerated, then it looks like it works initially; but if I add a space or change something else so it recompiles, then I get an error: "up is not defined".

File: src/RileyMaze/RMZUnit.js
Comment (ptomato) [inline src/RileyMaze/RMZUnit.js]: I'd suggest changing this to check whether the value is `'up'` or `'down'` and change the error message to say that the value must be up or down, since we don't allow just any string. (And we don't care if they use the string "up" or the variable "up")

File: src/RileyMaze/RMZUnit.js
Comment (ptomato) [inline src/RileyMaze/RMZUnit.js]: Currently there's a newline on line 1 of this code block, if you put a `\` after the backquote it would remove the newline and be more consistent with other code views.

File: src/RileyMaze/RMZUnit.js
Comment (ptomato) [inline src/RileyMaze/RMZUnit.js:222]: Up to you if you want to change this as part of this PR or later, but you could consider adding the `robotAAsset` and `robotBAsset` variables here even if they're not hooked up to the game yet.

File: src/RileyMaze/model.js
Comment (ptomato) [inline src/RileyMaze/model.js]: I think `rileyPosition` and `goalPosition` could be passed along to the game as part of the `levelCode` string, since the player defines them in the code view. There isn't a separate UI element where they can be set, so if you keep them inside `levelCode` then you wouldn't have to do any syncing between their values in the player's code block and in this model.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: The maximum X is 7.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js:62]: These checks should go before the previous checks, or they won't have an effect.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: Here I'd check if the unit is not just any string, but one of the allowed units.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js:15]: There was supposed to be an accompanying `ip` variable according to the API document, but I went ahead and just deleted that from the API document. It was secret anyway, and it's not actually needed.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js:12]: According to the API document this should be a JS Proxy so that we can catch invalid property accesses and flag them as unknown instructions. But I think this is actually fine for now, as long as we do the JS Proxy on the game side so that we know when to add a glitched instruction to the game UI.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js:20]: According to the API document, the instruction methods should throw an exception if the queue is already full, so we can display that to the player. (I saw that you are checking for >8 instructions below already, but throwing an exception here allows us to show the error dot at the right location.)

File: src/RileyMaze/userFunction.js
Comment (ptomato) [reply src/RileyMaze/userFunction.js]: Now that I think of it, we should probably put the `riley` object in its own scope (`INSTRUCTIONS_SCOPE`?) otherwise the Riley API will be available in the level editor too.

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: We have to move this somehow into a validation function that's separate for each user function panel. Otherwise the level editor will check if the Riley queue is full. (You can see this by adding a space or changing something else in the level code view so that it recompiles — you get "Instructions must have 8 moves")

**Review by ptomato — CHANGES_REQUESTED** [review]
Nice work. I went back to the API document and updated a few things after testing this branch and after the discussion about string constants the other day.

Most of the comments are minor and this is pretty much mergeable as is. If you want to split off the other ones into follow up tickets, so that we can get it merged tomorrow, that's fine with me.

**Comment by ptomato** [comment]
Oh, I almost forgot — before merging this should use the new name `com.endlessm.Sidetrack` instead of `com.endlessm.RileyMaze`

File: src/RileyMaze/userFunction.js
Comment (ptomato) [inline src/RileyMaze/userFunction.js]: The maximum X is now correct, but the maximum Y should be 4.

**Review by ptomato — APPROVED** [review]
As we discussed on Slack, I found only one thing that really should be addressed before merging. It's a minor fix, go ahead and merge this after fixing it. I'll leave a list of other things in the Phabricator ticket and you can decide what to do with them.

**Pull request merged**
